core/src/engine/impl/wait_list_light.cpp

```diff
@@ -1,81 +1,82 @@
 #include "wait_list_light.hpp"
 
-#include <engine/task/task_context.hpp>
-#include <userver/engine/sleep.hpp>
+#include <atomic>
+#include <cstddef>
+
 #include <userver/utils/assert.hpp>
-#include <userver/utils/fast_scope_guard.hpp>
+#include <userver/utils/underlying_value.hpp>
+
+#include <concurrent/intrusive_walkable_pool.hpp>
+#include <engine/task/task_context.hpp>
 
 USERVER_NAMESPACE_BEGIN
 
 namespace engine::impl {
 
-#ifndef NDEBUG
-WaitListLight::SingleUserGuard::SingleUserGuard(WaitListLight& wait_list)
-    : wait_list_(wait_list) {
-  auto& task = engine::current_task::GetCurrentTaskContext();
-  UASSERT(!wait_list_.owner_);
-  wait_list_.owner_ = &task;
-}
-
-WaitListLight::SingleUserGuard::~SingleUserGuard() {
-  auto& task = engine::current_task::GetCurrentTaskContext();
-  UASSERT(wait_list_.owner_ == &task);
-  wait_list_.owner_ = nullptr;
-}
-#endif
+struct WaitListLight::Impl final {
+  std::atomic<concurrent::impl::TaggedPtr<TaskContext>> waiter{nullptr};
+};
 
-WaitListLight::~WaitListLight() = default;
+WaitListLight::WaitListLight() noexcept : impl_() {}
 
-bool WaitListLight::IsEmpty() const { return waiting_; }
+WaitListLight::~WaitListLight() {
+  UASSERT_MSG(impl_->waiter.load().GetDataPtr() == nullptr,
+              "Someone is waiting on WaitListLight while it's being destroyed");
+}
 
-void WaitListLight::Append(
-    boost::intrusive_ptr<impl::TaskContext> context) noexcept {
+void WaitListLight::Append(boost::intrusive_ptr<TaskContext> context) noexcept {
   UASSERT(context);
   UASSERT(context->IsCurrent());
-  UASSERT(!waiting_);
-#ifndef NDEBUG
-  UASSERT(!owner_ || owner_ == context.get());
-#endif
-
+  UASSERT_MSG(!impl_->waiter.load(std::memory_order_relaxed).GetDataPtr(),
+              "Another coroutine is already waiting on this WaitListLight");
   LOG_TRACE() << "Appending, use_count=" << context->use_count();
-  waiting_ = context.detach();
+
+  // We can only fit 16-bit epoch into the atomic. This means that we allow no
+  // more than 2^16-1 coroutine sleep sessions during a single racy WakeupOne,
+  // as opposed to the regular 2^32-1 ones. Not a big deal, we use
+  // boost::lockfree queues with 16-bit tags anyway.
+  const auto cut_epoch = static_cast<std::uint16_t>(context->GetEpoch());
+
+  // Keep a reference logically stored in the WaitListLight to ensure that
+  // WakeupOne can complete safely in parallel with the waiting task being
+  // cancelled, Remove-d and stopped.
+  auto* const ctx = context.detach();
+
+  impl_->waiter.store({ctx, cut_epoch}, std::memory_order_seq_cst);
```

Comment:
I don't really get where these 16 bits fit, we might not have spare bits in the address in some configurations.